### DISCUSSION TITLE
Processing pipeline overhaul

### DISCUSSION THREAD

**Discussion by AbstractQbit** [OpenTabletDriver/OpenTabletDriver#685] [Ideas]
So right now the pipeline is quite rigid with a lot of parts hard-coded, like filter stages, interpolation, and output.

Instead what we can actually do is to make a gstreamer-like pipeline, where it is composed of different elements. Each element in the pipeline has a sink (event handler) as an input, and after processing anything it will pass data to the next element from it's source (event). The object that they pass along is a device report that they will modify along the way. 
That gives a ton of flexibility in how input might get processed, eliminating the need for those ugly filter stages. Also interpolators can be like any other element, just the receive/send part gets asynchronous

The only elements that don't have both I/O are Generators, who only produce input directly from a device, and Sinks that output to the system at the end of the pipeline.

Here is the rough example of how such pipeline might look like, keep in mind that basically nothing is hardcoded and each element needs to just implement receiving input and giving back output to the next thing
![image](https://user-images.githubusercontent.com/38468635/105175281-c0712680-5b34-11eb-8545-62f0927c0694.png)

**Comment by InfinityGhost** [comment]
So a base implementation of this would sorta be like this
```cs

// Base pipe interface
public interface IPipe<T>
{
    event Action<T> Send;
    void Recieve(T value);
}

public interface IDeviceReportPipe : IPipe<T>
{
}

// Filter example
public class Filter : IDeviceReportPipe
{
    public event Action<IDeviceReport> S;
    
    public void Recieve(IDeviceReport report)
    {
        // Modify report
        if (report is ITabletReport tabletReport)
            tabletReport.Position = report.Position * 0.9;
        Send?.Invoke(report);
    }
}

// Interp example
public class Interpolator : IDeviceReportPipe
{
    private IDeviceReport state;

    public event Action<IDeviceReport> Send;

    public void Recieve(IDeviceReport report)
    {
        state = report;
    }

    // started in some way by the ctor
    private void mainLoop()
    {
        Send?.Invoke(state);
    }
}
```

Current issues I can see with this is that there is no way to set the position in the pipeline, but we'd need some sort of pipeline manager anyway in order to hook each one together.

**Reply by AbstractQbit** [reply]
Yeah, that's what it should look like pretty much. Maybe we should stick to GStreamer's naming (PipelineElement, source, sink), as it seems a bit more clear? IPipe sounds like something to do with IPC, which we also have in the project.

**Reply by InfinityGhost** [reply]
>  IPipe sounds like something to do with IPC

Yeah good point, we have enough pipes to deal with as it is lol

**Comment by InfinityGhost** [comment]
About time to look back at this, anything we should change about the process or naming schemes?
